src/concat.js

```diff
@@ -25,12 +25,12 @@ var _isArray = require('./internal/_isArray');
  *      R.concat([4, 5, 6], [1, 2, 3]); //=> [4, 5, 6, 1, 2, 3]
  *      R.concat('ABC', 'DEF'); // 'ABCDEF'
  */
-module.exports = _curry2(function(set1, set2) {
-  if (_isArray(set2)) {
-    return _concat(set1, set2);
-  } else if (_hasMethod('concat', set1)) {
-    return set1.concat(set2);
+module.exports = _curry2(function(a, b) {
+  if (_isArray(a) && _isArray(b)) {
+    return _concat(a, b);
+  } else if (_hasMethod('concat', a)) {
```

Comment:
This implementation passes all the tests on master:

```javascript
module.exports = flip(invoker(1, 'concat'));
```

So if all we want to do is handle valid input we needn't do anything more.

The fact that we instead choose to include a number of checks suggests that we're interested in validating the input. I just don't understand your algorithm for determining which *detectable* contract breaches to detect and which to ignore.

```javascript
R.concat(null, []);  // currently detected
R.concat([], null);  // not currently detected; this pull request adds detection
R.concat(null, '');  // currently detected
R.concat('', null);  // not currently detected; this pull request does not add detection
R.concat([], '');    // currently detected
R.concat('', []);    // not currently detected; this pull request does not add detection
```

I argue that we should either detect and report contract breaches consistently or remove the checks entirely. Our current checks—and the check added in this pull request—strike me as half measures.